### PR TITLE
Integrate storage

### PR DESCRIPTION
**Pull request by ebiiii**
- [x] storage configuration need an access token of a user, which cannot (easily) be generated from the docker-compose. Maybe to be added in makefile?
- [x] lfs doesn't work if enabled on gitlab side (and also the storage service needs a bucket to exists). 
- [x] swagger definition is outdated on the storage service.
- [x] document steps needed to use storage with projects

# Setup to integrate storage:

## Getting credentials for testing
- Renga configuration of Keycloak sets a 5 minutes expiration on token. You may want to change it to some longer lifespan for manual testing. In a deployed environment, refresh token and offline tokens should be used by the client.
- To get a JWT from keycloak (using the default renga configuration):
 ```curl -sS -X POST -d "client_id=renga-services&client_secret=d7f63e67-e74d-4935-bc4b-7803f2b995e8&grant_type=client_credentials&redirect_uri=http://localhost" "keycloak.renga.build:8080/auth/realms/Renga/protocol/openid-connect/token"```

## LFS only with direct gitlab git repository connection (existing repository in gitlab)
  - disable LFS on the gitlab project (it seems only admin can do it from the admin view of projects)
  - once the local git repository is cloned, move into it and set the credentials for LFS: ```git config http.extraheader "authorization: bearer <insert the JWT here>"```
  - create a bucket for the LFS objects ( you can also choose to have a single bucket for all projects ): ```curl -X POST "http://renga.build/api/storage/repo" -H "accept: application/json" -H "authorization: bearer <insert the JWT here>" -H "content-type: application/json" -d '{ "uuid": "7fe0dd91-dbd4-4d1c-b6cc-c1837b66b31d", "description": "a repository", "path": "lfs-data", "backend": {}, "lfs_store": "7fe0dd91-dbd4-4d1c-b6cc-c1837b66b31d"}' ``` The UUID can be chosen randomly, but must be the same in both uuid and lfs_store fields.
  - setup LFS (```git lfs install``` and add some tracked files ```git lfs track ...```)
  - tell LFS to use the storage service and the given bucket: ```git config -f .lfsconfig lfs.url http://renga.build/api/storage/repo/7fe0dd91-dbd4-4d1c-b6cc-c1837b66b31d.git/info/lfs``` 
  - You can then work normally with git and LFS 
  - to pull into a new repository or on a new machine, download from LFS should be disabled first (for example with ```GIT_LFS_SKIP_SMUDGE=1 git clone ...```. Then the JWT added like in the 3rd step, and only then it is possible to pull from lfs (```git lfs pull```).

## LFS with storage service managed repository
 - create a git repository ( by default a bucket is created for the LFS objects per repository ): ```curl -X POST "http://renga.build/api/storage/repo" -H "accept: application/json" -H "authorization: bearer <insert the JWT here>" -H "content-type: application/json" -d '{ "uuid": "7fe0dd91-dbd4-4d1c-b6cc-c1837b66b31d", "description": "a git repository", "path": "my-project", "backend": {"name": "gitlab"}}' ``` The UUID can be chosen randomly, it is also possible to specify an existing bucket for the lfs_store.
  - clone the repository: ```git -c http.extraheader="authorization: bearer <insert JWT>" clone http://renga.build/api/storage/repo/7fe0dd91-dbd4-4d1c-b6cc-c1837b66b31d.git my-repository```
  - move into it and set the credentials for git: ```git config http.extraheader "authorization: bearer <insert the JWT here>"```
  - setup LFS (```git lfs install``` and add some tracked files ```git lfs track ...```)
  - You can then work normally with git and LFS

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -187,6 +192,16 @@ unregister-gitlab-oauth-applications:
...
 
+register-gitlab-user-token: unregister-gitlab-user-token
+	@$(DOCKER_COMPOSE_ENV) docker-compose exec gitlab /opt/gitlab/bin/gitlab-psql \
+		-h /var/opt/gitlab/postgresql -d gitlabhq_production \
+		-c "INSERT INTO personal_access_tokens ( id, user_id, token, name, revoked, expires_at, created_at, updated_at, scopes, impersonation) VALUES ( '1', '1', '$(GITLAB_TOKEN)', 'storage token', 'f', NULL, '2018-02-26 10:46:16.602039', '2018-02-26 10:46:16.602039', E'--- \n- api\n- read_user\n- sudo\n- read_registry', 'f');"
+
+unregister-gitlab-user-token:
+	@$(DOCKER_COMPOSE_ENV) docker-compose exec gitlab /opt/gitlab/bin/gitlab-psql \
+		-h /var/opt/gitlab/postgresql -d gitlabhq_production \
+		-c "DELETE FROM personal_access_tokens WHERE id='1'"
```
Comment (jirikuncar) [inline Makefile]: This is not a very safe way!

File: docker-compose.yml [hunk 2]
```diff
@@ -148,24 +147,54 @@ services:
...
-#     labels:
-#       traefik.enable: "true"
-#       traefik.frontend.rule: PathPrefix:/api/storage
-
+  storage:
+    image: ${DOCKER_REPOSITORY}renga-storage:${PLATFORM_VERSION}
+    environment:
+      AUTHORIZATION_CLIENT_SECRET: $AUTHORIZATION_CLIENT_SECRET
+      PLAY_APPLICATION_SECRET: $PLAY_APPLICATION_SECRET
+      RENGA_SERVICES_AUTHORIZATION_URL: ${RENGA_ENDPOINT}/auth/realms/Renga/protocol/openid-connect/auth
```
Comment (jirikuncar) [inline docker-compose.yml]: this might be `http://keycloak.${...}/realms...`

Comment (ebiiii) [reply docker-compose.yml]: This is consistent with line 63. Further changes to using domain name could be done in a separate PR.

Comment (rokroskar) [reply docker-compose.yml]: `docker-compose` has changed in the meantime. Unless we choose to revert #133 with #138 this is inconsistent. In any case, you can use `${KEYCLOAK_URL}/auth/...` here for the time being and I think you're ok with both options.

File: Makefile [hunk 3]
```diff
@@ -187,6 +192,16 @@ unregister-gitlab-oauth-applications:
 		-h /var/opt/gitlab/postgresql -d gitlabhq_production \
 		-c "DELETE FROM oauth_applications WHERE uid='renga-ui'"
 
+register-gitlab-user-token: unregister-gitlab-user-token
+	@$(DOCKER_COMPOSE_ENV) docker-compose exec gitlab /opt/gitlab/bin/gitlab-psql \
+		-h /var/opt/gitlab/postgresql -d gitlabhq_production \
+		-c "INSERT INTO personal_access_tokens ( user_id, token, name, revoked, expires_at, created_at, updated_at, scopes, impersonation) VALUES ( '1', '$(GITLAB_TOKEN)', 'managed storage token', 'f', NULL, '2018-02-26 10:46:16.602039', '2018-02-26 10:46:16.602039', E'--- \n- api\n- read_user\n- sudo\n- read_registry', 'f');"
```
Comment (rokroskar) [inline Makefile]: use current time?

Comment (ebiiii) [reply Makefile]: done

File: docker-compose.yml (hunk 2 above)
Comment (ebiiii) [reply docker-compose.yml]: Changed it. Gitlab runner will probably also need this change in its configuration. https://github.com/SwissDataScienceCenter/renga/pull/128/files#diff-4e5e90c6228fd48698d074241c2ba760R148

Comment (rokroskar) [reply docker-compose.yml]: ah yes, absolutely, good catch.

**Comment by rokroskar** [comment]
@ebiiii which branch of `renga-storage` should we actually look at/build to try out this PR?

**Comment by rokroskar** [comment]
can you give an example of how to use the storage service for LFS? I tried this:

```
git config -f .lfsconfig lfs.url http://renga.local/api/storage
```

without much success

File: Makefile [hunk 4]
```diff
@@ -191,22 +191,22 @@ services/gitlab/%:
...
 	@$(DOCKER_COMPOSE_ENV) docker-compose exec gitlab /opt/gitlab/bin/gitlab-psql \
 		-h /var/opt/gitlab/postgresql -d gitlabhq_production \
-		-c "DELETE FROM oauth_applications WHERE uid='renga-ui'" >& /dev/null
+		-c "DELETE FROM oauth_applications WHERE uid='renga-ui'" > /dev/null 2>&1
 
 register-gitlab-user-token: unregister-gitlab-user-token
 	@$(DOCKER_COMPOSE_ENV) docker-compose exec gitlab /opt/gitlab/bin/gitlab-psql \
 		-h /var/opt/gitlab/postgresql -d gitlabhq_production \
-		-c "INSERT INTO personal_access_tokens ( user_id, token, name, revoked, expires_at, created_at, updated_at, scopes, impersonation) VALUES ( '1', '$(GITLAB_TOKEN)', 'managed storage token', 'f', NULL, '2018-02-26 10:46:16.602039', '2018-02-26 10:46:16.602039', E'--- \n- api\n- read_user\n- sudo\n- read_registry', 'f');"
+		-c "INSERT INTO personal_access_tokens ( user_id, token, name, revoked, expires_at, created_at, updated_at, scopes, impersonation) VALUES ( '1', '$(GITLAB_TOKEN)', 'managed storage token', 'f', NULL, '$(shell date +'%Y-%m-%d %H:%M:%S')', '$(shell date +'%Y-%m-%d %H:%M:%S')', E'--- \n- api\n- read_user\n- sudo\n- read_registry', 'f');" > /dev/null 2>&1
```
Comment (jirikuncar) [inline Makefile]: what about using postgres `now()` function?

**Comment by ebiiii** [comment]
@rokroskar yep it is still from the remove_graph branch. Should be soon merged to development.
Regarding the urls, I forgot to mention that you need to create the bucket first (```curl -X POST "http://renga.local/api/storage/repo" -H  "accept: application/json" -H  "authorization: bearer " -H  "content-type: application/json" -d "{  \"uuid\": \"7fe0dd91-dbd4-4d1c-b6cc-c1837b66b31d\",  \"description\": \"a repository\",  \"path\": \"lfs-data\",  \"backend\": {},  \"lfs_store\": \"7fe0dd91-dbd4-4d1c-b6cc-c1837b66b31d\"}"``` you can use any random uuid as long as you put twice the same.) Then the url for LFS is  http://renga.local/api/storage/repo/7fe0dd91-dbd4-4d1c-b6cc-c1837b66b31d.git/info/lfs
I haven't tested it in this form yet.

**Comment by rokroskar** [comment]
thanks @ebiiii -- would be great to have a full, tested, working example. I'm a bit lost as to the envisioned flow here -- would we create a new bucket for each repo automatically for the users (i.e. the UI or the CLI would create one)? Or would multiple repos go into the same bucket? 

I just tried this with a sample repo I'm playing with and it seems to stall when I push. Following your suggestion above, I made the repo (bucket?) via the swagger UI with this JSON:

```
{
  "uuid": "7fe0dd91-dbd4-4d1c-b6cc-c1837b66b31d",
  "description": "test-repo",
  "path": "lfs-data",
  "backend": {"name":"local"},
  "lfs_store": "7fe0dd91-dbd4-4d1c-b6cc-c1837b66b31d"
}
```

I push to a repo that includes some LFS-tracked files, I get asked for username/password and then nothing happens. 

(a related question: why should the client be creating the uuids?)

**Comment by rokroskar** [comment]
@ebiiii could you please document all the steps needed to (as a first step) use the LFS server with renga projects? If there are additional initializations that have to take place for each project (e.g. creating a bucket) then we also need to sync what is being done in the CLI.

**Comment by erbou** [comment]
I can create a repo, clone, add a file and commit/push. The readme needs to be updated to use renga.build instead of renga.local.

File: Makefile [hunk 5]
```diff
@@ -66,7 +66,9 @@ DOCKER_COMPOSE_ENV=\
 	PLATFORM_VERSION=$(PLATFORM_VERSION) \
 	RENGA_ENDPOINT=$(RENGA_ENDPOINT) \
 	RENGA_UI_URL=$(RENGA_UI_URL) \
-	PLAY_APPLICATION_SECRET=$(PLAY_APPLICATION_SECRET)
+	PLAY_APPLICATION_SECRET=$(PLAY_APPLICATION_SECRET) \
+	GITLAB_TOKEN=$(GITLAB_TOKEN)
```
Comment (jirikuncar) [inline Makefile]: please, can you sort the list of env variables?

**Comment by leafty** [comment]
git `http.extraheader` was imtroduces in git 2.9.0. There will probably be issues with folks on ubuntu (git 2.8.4, >2 years old :/)

**Comment by 3C111** [comment]
It works for me!

**Pull request merged**